### ISSUE TITLE
cache is not cleaned

### ISSUE THREAD

**Issue by hachreak** [wkeeling/selenium-wire#35]
Hi everybody,
I have an issue about the stored cache.
The software doesn't clean it.
In my logs I have a lot of `ConnectionResetError` that doesn't crash the application completely but probably could explain why is not cleaning the cache.

This is an example of exception:
```
Traceback (most recent call last):
  File "/usr/lib/python3.5/socketserver.py", line 625, in process_request_thread
    self.finish_request(request, client_address)
  File "/usr/lib/python3.5/socketserver.py", line 354, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/home/hachreak/projects/scraper/lib/python3.5/site-packages/seleniumwire/proxy/proxy2.py", line 65, in __init__
    super().__init__(*args, **kwargs)
  File "/usr/lib/python3.5/socketserver.py", line 681, in __init__
    self.handle()
  File "/usr/lib/python3.5/http/server.py", line 424, in handle
    self.handle_one_request()
  File "/usr/lib/python3.5/http/server.py", line 390, in handle_one_request
    self.raw_requestline = self.rfile.readline(65537)
  File "/usr/lib/python3.5/socket.py", line 576, in readinto
    return self._sock.recv_into(b)
  File "/usr/lib/python3.5/ssl.py", line 937, in recv_into
    return self.read(nbytes, buffer)
  File "/usr/lib/python3.5/ssl.py", line 799, in read
    return self._sslobj.read(len, buffer)
  File "/usr/lib/python3.5/ssl.py", line 583, in read
    v = self._sslobj.read(len, buffer)
ConnectionResetError: [Errno 104] Connection reset by peer
```

I suspect that it's crashing the thread with the proxy that skip the [cleaning](https://github.com/wkeeling/selenium-wire/blob/master/seleniumwire/proxy/server.py#L65).
Do you have any suggestion how it's possible to fix this behavior?
Thanks a lot! :smile:

**Comment by wkeeling** [comment]
@hachreak Thanks for raising this.

This error is often seen when the browser quits and selenium wire's proxy server receives a signal to shut down. The shut down process is not instantaneous however, and sometimes requests that were in-progress inside the proxy will be sent back to the browser, which is now closed, triggering the "Connection reset by peer" message. This is normally harmless, and the message is thus logged at debug level.

That said, we should definitely investigate whether that error has the side effect of preventing the stored requests from being cleared. I'll check that out and report back here.

**Comment by wkeeling** [comment]
@hachreak Do you have a simple example of a test that exhibits this issue? I've not been able to reproduce the issue so far, so it would be good to compare how you've set your test up/configured selenium-wire.

**Comment by hachreak** [comment]
Hi @wkeeling , thanks for looking for the issue.
I would say that, in normal conditions, the software does his job and it doesn't raise this exception.
I don't know a reproducible way to generate a `ConnectionResetError`.
Usually It happen to me when I try to open a lot of pages on a selenium driver.
Maybe you can "artificially" raise the exception inside the `proxy2.py` to simulate the issue and see what's happening. WDYT?

**Comment by wkeeling** [comment]
@hachreak - ok no problem. I'll see if I can force the condition by artificially raising that exception as you suggest.

**Comment by wkeeling** [comment]
@hachreak - struggling to reproduce the issue even by raising the exception manually from different places in the code. Out of curiosity, are you explicitly closing the browser at the end of the test? If so, do you use `driver.quit()` or `driver.close()`, or neither?

**Comment by hachreak** [comment]
This is an example where I try to "hydrate" [twitter post ids](https://github.com/hachreak/scraper/blob/master/scraper/cli/main.py#L88).
I'm loading the driver and the `with` should then automatically call the [quit](https://github.com/hachreak/scraper/blob/master/scraper/driver.py#L56) before exit.
Do you think I'm doing something wrong? :smile: 

I have a question: what can happen if I open many scraper in parallel?

**Comment by wkeeling** [comment]
@hachreak your code looks fine. I think the issue is that Selenium Wire is not thorough enough in cleaning up after itself. It clears out the cached request folders each time you call `driver.quit()` - but it leaves the parent folders behind expecting an exit handler to clean those up. If that handler doesn't fire then the folders will remain.

I've modified `driver.quit()` to properly clean up the whole directory tree which should resolve the issue I hope.

**Comment by wkeeling** [comment]
Regarding the errors you're seeing, I don't think are related to the folder deletion issue and are due to Firefox disappearing before Selenium Wire has been able to read all the data from the socket. These errors are normally harmless, although they look alarming (other people have reported similar connection related tracebacks). 

I am planning to suppress connection errors by default - and log them at DEBUG level. These can be unsuppressed via a config option.

**Comment by wkeeling** [comment]
Regarding your last question: _what can happen if I open many scraper in parallel?_ Selenium Wire should be able to handle this, because each local storage area is unique per test instance. Also, the embedded proxy server started by Selenium Wire will always start on a unique port number - so multiple instances can exist at the same time.

Obviously if you hit any issues with parallel execution then feel free to open a new issue.

**Comment by wkeeling** [comment]
Cache cleanup and error suppression now available in version 1.0.4
